File: 0452-minimum-number-of-arrows-to-burst-balloons/0452-minimum-number-of-arrows-to-burst-balloons.py
class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        if not points:
            return 0
    
        # Sort by end coordinate (sweep left to right)
        points.sort(key=lambda x: x[1])
        
        arrows = 1
        arrow_pos = points[0][1]  # Shoot at end of first balloon
        
        for start, end in points[1:]:
            # Current balloon not covered by last arrow
            if start > arrow_pos:
                arrows += 1
                arrow_pos = end  # Shoot at end of this balloon
        
        return arrows
        

File: 0452-minimum-number-of-arrows-to-burst-balloons/0452-minimum-number-of-arrows-to-burst-balloons.java
class Solution {
    public int findMinArrowShots(int[][] points) {
        List<int[]> pList = new ArrayList<>();

        for (int[] point : points) {
            pList.add(new int[]{point[0], point[1]});
        }

        Collections.sort(pList, (a, b) -> a[1] == b[1] ? Integer.compare(a[0], b[0]) : Integer.compare(a[1], b[1]));
        int prev = (int) pList.get(0)[1], val = 1;
        for (int i = 1; i<pList.size(); i++) {
            int[] point = pList.get(i);
            int start = point[0];
            int end = point[1];

            if (start > prev) {
                val++;
                prev = end;
            }
        }

        return val;
    }
}